### ISSUE TITLE
twistd should support external log rotation facilities

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3363]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#3363|
|Type|enhancement|
|Created|2008-07-22 15:32:49Z|

This basically means that it should be possible to get twistd to handle a signal to re-open its logfiles.

Attachments:

* [ticket-3363.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a75/a753b79cc20aaf53061b3fd2d55af85b9137f415/b481873b600f83c4f9a2bfba258caf7274070b33.patch) (4475 bytes) - added by henninge on 2010-07-14 14:35:57Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3363 3363
type__enhancement enhancement
reporter__radix radix
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__reopened reopened
resolution__ 
component__core core
keywords__twistd twistd
time__1216740769000000 1216740769000000
changetime__1280242851000000 1280242851000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="jasonjwwilliams's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|jasonjwwilliams commented|
|-|-|

Isn't this implemented already? Issuing SIGUSR1 forces a reload?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

Yes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `reopened`|
|-|-|

Actually, as it turns out, no.  SIGUSR1 forces a rotation, not a re-open.  This is not quite the same.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

[#4557](https://github.com/twisted/twisted/issues/4557) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="henninge's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|henninge commented|
|-|-|

The patch I submitted adds an "--nologrotation" option that deactivates log rotation and makes twistd reopen its log file on SIGUSR1 instead. Without this option the old behavior is retained.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

There was some discussion of this issue on IRC.  One factor to consider is the proliferation of logging configuration options accepted by `twistd`.  So far, we've been avoiding introducing new options because we have no general plan for `twistd`.  I'd personally like to avoid having ten (or twenty or fifty) command line flags just for configuring `LogFileObserver`.  Not just because this kind of mess makes it hard for users to find what they want, but also because `LogFileObserver` is just one possible observer.  It would be great if we could come up with a plan which allowed arbitrary configuration of a (or several) log observer without having to extend `twistd` for every option.

I'm not sure if I should leave this ticket in review or not, so I'm going to for now.  It's possible that in a few days I'll come back and say something more. :)

Also, I went through and marked all the twistd-related issues that I could find with the "twistd" keyword, so you can now look at them all with this easy query:

  http://twistedmatrix.com/trac/query?status=assigned&status=new&status=reopened&order=priority&col=id&col=summary&col=status&col=type&col=priority&col=milestone&col=component&keywords=~twistd

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

I think for now `--nologrotation` may be the way to go for now, since it is a small fix.  The point about a long-term plan is well-taken though.

Two other areas which have faced similar option proliferation are `ICredentialsChecker` construction, for which we now have strcred, and `IListeningPort` construction, for which we now have `strports` (and [ticket:4473 soon], I hope, "`strpoints`").

The general pattern of the solution is simple: instead of `--foo-a` `--foo-b` `--foo-c`, we have a (pluggable) `--thunk foo:a,b,c` option.  `strports` and `endpoints` are both fairly simple, and we could probably cram most of a log configuration system into something similar.

If we did this for logging, not only could we support different loggers, we could also support _multiple_ loggers of the same type.

This is the basic idea behind the way the 'logging' module configures itself; you can set up any number of loggers and configure them in various ways.  Unfortunately loggers are probably going to push the boundaries of what's comfortable to type on the command line.  Take, for example, [the "simple" example logging.conf file](http://docs.python.org/library/logging.html#configuring-logging) in the "configuring logging" section of the logging module's documentation.  Of course, we could address this with `--log-observer logging:/path/to/my/logging.conf`.

As a general high-level plan for all of these systems that take command-line options to construct something, we might want to consider having a 'python' plugin which execfiles a Python source file and pulls a variable out of it (or maybe even just evaluates a Python statement provided directly on the command line, for simple scripts) to satisfy whatever interface is required.  This would serve a dual purpose.  First, it would give you a way to do quick-and-dirty customization via something like a .tac file, but for small portions of the twistd application service structure instead of needing to build the whole thing yourself.  Second, it would provide a straightforward way to integrate things for which there are too many options to stick into a single command-line string in a consistent way.

To sum up, this particular use-case would be addressed with `--log-observer simple:/path/to/my/log`.  The behavior we've got right now could be requested with `--log-observer rotating:/path/to/my/log`.  Each of these would still want to grab `SIGUSR1`, so we may want to add a cooperative system for listening to that signal, or for requesting a signal to listen to, and then logging the allocation of each signal  ("send SIGUSR1 to rotate the log at ...", "send SIGUSR2 to close the log at ...").  Even POSIX guarantees us up to 8 signals between `SIGRTMIN` and `SIGRTMAX`.

Should we file a new ticket for doing this for logging?

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_11"></a>|@khorn commented|
|-|-|

Would it be possible to have something like apache's -D flag, where you have one option that can be used multiple times and that takes key/value pairs?

so if -x is the "override logging flag", you might do:
```
-xoption1=value1 -xoption2=value2
```

or something like that.

Something like that could even be more general than just logging...you could pass these values to the .tac file or plugin or whatever and let it decide what to do.

I don't know whether this is a _good_ idea, mind you...

**Comment by twisted-trac** [comment]
|<img alt="henninge's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|henninge commented|
|-|-|

I solved this for now by defining my own log observer and applying it in the tac file. Luckily twistd checks for a component in the application that provides ILogObserver.
http://paste.ubuntu.com/464982/

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

Many comments since the keyword review was added, should this ticket still be in review or..

**Comment by twisted-trac** [comment]
|<img alt="henninge's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|henninge commented|
|-|-|

No, I think not.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
